### PR TITLE
Add new product to listing

### PR DESCRIPTION
**Pull request by dmytrokulak**
[Card #72](https://trello.com/c/Rx4PrA8w/72-make-the-checkmark-button-on-edit-page-to-add-a-new-product-in-the-listing)

Adds a feature to save a new product when clicking on checkmark on edit page.

### CODE REVIEW

**Comment by cerrmor** [comment]
I like the add product feature, it seems to work well however, minor issue I found is it is not correctly displaying the number of days left when a new product is added and it doesn't register a unique key to the element so the console throws a warning. 

I think however that we can merge this with the main branch

**Comment by dmytrokulak** [comment]
> I like the add product feature, it seems to work well however, minor issue I found is it is not correctly displaying the number of days left when a new product is added and it doesn't register a unique key to the element so the console throws a warning.
> 
> I think however that we can merge this with the main branch

@cerrmor Good catch, thanks! I'll fix that. I don't think it is something addressed on acceptance testing PR.

File: src/Products/sources/TestSource.js [hunk 1]
```diff
@@ -2,8 +2,45 @@ import testData from './test-data.json';
...
+    });
+
     return testData[name];
-  }
-}
+  },
+
+  write(name, entity, id) {
+    //try remove old version if exists
+    if (id) {
```
Comment (brurez) [inline src/Products/sources/TestSource.js:15]: If we have time to use LocalStorage as `DataSource` or if we decide to implement a backend service to persist the products, we will have to copy and paste this logic to `src/Products/sources/LocalStorageSource.js` for example. What do you think to move this logic to `/src/Products/index.js`? This way we leave the data sources with just the responsibility to persist the product, nothing else.

**Review by brurez — CHANGES_REQUESTED** [review]

File: src/Products/sources/TestSource.js [hunk 2]
```diff
@@ -2,8 +2,45 @@ import testData from './test-data.json';
...
+      entity.daysLeft = entity.coolingPeriod;
+    }
+    //persist entity
+    testData[name].push(entity);
+  },
+};
+
+//createdAt: string in ISO format
+const calculateDaysLeft = (createdAt, coolingPeriod) => {
+  debugger
```
[the remark this answers is not in the document]
Comment (dmytrokulak) [reply src/Products/sources/TestSource.js]: thanks,. my bad. 
7c90f2f

File: src/Products/sources/TestSource.js (hunk 1 above)
Comment (dmytrokulak) [reply src/Products/sources/TestSource.js:10]: Good point. I've moved some logic to index.js. Though, I think, what concerns id generation - it's data source responsibility. A lot of Web APIs and underlying databases return autogenerated ids, so I'd leave it here.  
eb77c82

**Review by brurez — CHANGES_REQUESTED** [review]
Just update `Products/index.test.js` with the code below and we are good to go:
```
import { Products, TestSource } from "./index";

describe("Products", () => {
  describe(".getAll", () => {
    const expectedProducts = [
      {
        id: 1,
        name: "First product",
        likeCount: 3,
        linkToBuy: "www.lorem.ipsum",
        reasonToBuy: "dolor",
        reminderPeriod: 7,
        coolingPeriod: 30,
        createdAt: "2021-01-01T18:00:00.000Z",
        daysLeft: 0,
      },
      {
        id: 2,
        name: "Second product",
        likeCount: 1,
        linkToBuy: "www.lorem.ipsum",
        reasonToBuy: "dolor",
        reminderPeriod: 7,
        coolingPeriod: 30,
        createdAt: "2021-01-10T18:00:00.000Z",
        daysLeft: 0,
      },
      {
        id: 3,
        name: "Third product",
        likeCount: 5,
        linkToBuy: "www.lorem.ipsum",
        reasonToBuy: "dolor",
        reminderPeriod: 7,
        coolingPeriod: 30,
        createdAt: "2021-02-14T18:00:00.000Z",
        daysLeft: 24,
      },
    ];
    test("if returns products", () => {
      const products = new Products(TestSource);
      const allProducts = products.getAll();
      expect(allProducts).toEqual(expectedProducts);
    });
  });
});
```

**Comment by dmytrokulak** [comment]
> Just update `Products/index.test.js` with the code below and we are good to go:
> 
> ```
> import { Products, TestSource } from "./index";
> 
> describe("Products", () => {
>   describe(".getAll", () => {
>     const expectedProducts = [
>       {
>         id: 1,
>         name: "First product",
>         likeCount: 3,
>         linkToBuy: "www.lorem.ipsum",
>         reasonToBuy: "dolor",
>         reminderPeriod: 7,
>         coolingPeriod: 30,
>         createdAt: "2021-01-01T18:00:00.000Z",
>         daysLeft: 0,
>       },
>       {
>         id: 2,
>         name: "Second product",
>         likeCount: 1,
>         linkToBuy: "www.lorem.ipsum",
>         reasonToBuy: "dolor",
>         reminderPeriod: 7,
>         coolingPeriod: 30,
>         createdAt: "2021-01-10T18:00:00.000Z",
>         daysLeft: 0,
>       },
>       {
>         id: 3,
>         name: "Third product",
>         likeCount: 5,
>         linkToBuy: "www.lorem.ipsum",
>         reasonToBuy: "dolor",
>         reminderPeriod: 7,
>         coolingPeriod: 30,
>         createdAt: "2021-02-14T18:00:00.000Z",
>         daysLeft: 24,
>       },
>     ];
>     test("if returns products", () => {
>       const products = new Products(TestSource);
>       const allProducts = products.getAll();
>       expect(allProducts).toEqual(expectedProducts);
>     });
>   });
> });
> ```

@brurez  This won't work because daysLeft is calculated based on the current date.  I'll have to inject some DateService into Products class so I could mock that DateService in the tests. I can try.

**Review by brurez — APPROVED** [review]

**Pull request merged**